### PR TITLE
Add move_toward_angle and angle_difference methods

### LINKED ISSUE #2074: Add angular analogs of `move_toward()`
### Describe the project you are working on
A top-down game, with automation of combatants through player made behavior trees.

### Describe the problem or limitation you are having in your project
I'm making movement code for my top down characters, and I keep finding that Vector2 `move_toward()`
function is handy when moving kinematic bodies towards a point, but making them look at a point smoothly is hard.
look_at snaps something instantly, which isn't desired in this case.
Implementing and getting the wrapping logic took a chunk of brain power, to make sure it wasn't going to have subtle bugs with larger rotations, or negative ones.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Adding two functions to the core GDScript namespace, with the following WIP names.
`rotate_toward(angle_a, angle_b, amount)`:
This takes three angles as floats, and returns angle_a, rotated towards angle_b, by amount, or less if angle_b can be reach in less than amount. This would perform wrapping around TAU, and moving along the shortest arc between a and b.

`rotate_toward_degrees` would be as above, but wrapping around 360 degrees instead, and could potentially be omitted.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Consider a topdown games turret. Having it move smoothly to look at a target could thus be.

```
var target_angle = (target_position - global_position).angle()
$TurretBase.global_rotation = rotate_toward($TurretBase.global_rotation, target_angle, turret_turn_speed)
```
Similar example could be used for top down vehicles, or burrowing/flying platformer enemies.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
It wouldn't be hard to write around with scripts, but it would be easy to get subtly wrong, and would potentially have to be repeated in many file throughout a project.,

### Is there a reason why this should be core and not an add-on in the asset library?
Importing an add-on for this would be comparable to the amount of work required to do it in the first place. This aims to solve a papercut, like [#42976 ](https://github.com/godotengine/godot/pull/42976)did.

*Bugsquad edit, search terms: move_towards_angle*

### PR DESCRIPTION
**Pull request by ettiSurreal**
Closes https://github.com/godotengine/godot-proposals/issues/2074

Adds:

- `angle_difference`, which finds a value between 0 and TAU, finding the shortest path.
- `move_toward_angle`, moves towards an angle by an increment, wrapping around TAU and also finding the shortest path, similar to `lerp_angle`.
- modifies `lerp_angle` code to avoid repetition.

Implementation based on https://github.com/godotengine/godot-proposals/issues/5596#issuecomment-1282661563 and https://github.com/godotengine/godot/pull/37078.

Tested on a custom build, everything seems to be working as it should.

This is my first time making a PR, so let me know what to do if i messed something up!!

### CODE REVIEW

**Comment by ettiSurreal** [comment]
Since 4.0 is in feature freeze and since .NET 6 isn't mentioned in the compiling docs i assume building the C# version isn't fully documented yet, i'll most likely just wait for 4.0 to go in stable before updating the PR again.

**Comment by adamscoble** [comment]
@ettiSurreal Oh, this is great to see! Though to stay in line with other functions, should `angle_difference` just be called `angle`? 

Also one other consideration is that in the `move_toward_angle` method, I offset the start value instead of the end, so that the end value is actually reached. Most people will just be testing for the angle anyway, but I do like that idea that the target value is the result of the method, rather than a possible multiple of TAU from it?

Here's the [plugin](https://godotengine.org/asset-library/asset/1549) I made that implements it in case you're curious!

**Comment by PrecisionRender** [comment]
To make things easy to track, make sure to [squash your commits](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase) into a single commit. 👍

**Comment by Zireael07** [comment]
Bump. What's the state of this? Can this go into 4.0 or is this something that's for 4.1?

**Comment by Calinou** [comment]
> Bump. What's the state of this? Can this go into 4.0 or is this something that's for 4.1?

Any non-critical change is for 4.1 at the earliest now.

**Comment by AThousandShips** [comment]
> I hope that's not big of an issue, sorry.

You will have to squash before this can be merged once approved, hopefully you can get it to work

**Comment by AThousandShips** [comment]
It's hard, so no worries, I've struggled at the start too!

**Comment by kleonc** [comment]
I'll try to rebase. :upside_down_face:

Generally the PR looks ok, I'd say the docs need some more clarification. But I'm not fully sure what's the desired behavior in the first place. E.g. for the current implementation:

```gdscript
move_toward_angle(deg_to_rad(3600), deg_to_rad(90), deg_to_rad(45)) # Returns: deg_to_rad(3645)
move_toward_angle(deg_to_rad(90), deg_to_rad(3600), deg_to_rad(45)) # Returns: deg_to_rad(45)

move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(3600)) # Returns: deg_to_rad(90)
move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(-3600)) # Returns: deg_to_rad(-3600)
```

Is this expected/desired? Or do we maybe want to somehow limit the results to some range (like `[0, TAU)`)? :thinking:

**Comment by ettiSurreal** [comment]
> I'll try to rebase. 🙃 _Edit: Rebased. I think I haven't changed anything besides the commit name._
> 
Thanks for the help!
> Generally the PR looks ok, I'd say the docs need some more clarification. But I'm not fully sure what's the desired behavior in the first place. E.g. for the current implementation:
> 
> ```gdscript
> move_toward_angle(deg_to_rad(3600), deg_to_rad(90), deg_to_rad(45)) # Returns: deg_to_rad(3645)
> move_toward_angle(deg_to_rad(90), deg_to_rad(3600), deg_to_rad(45)) # Returns: deg_to_rad(45)
> 
> move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(3600)) # Returns: deg_to_rad(90)
> move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(-3600)) # Returns: deg_to_rad(-3600)
> ```
> 
> Is this expected/desired? Or do we maybe want to somehow limit the results to some range (like `[0, TAU)`)? 🤔

As for the docs i am not the best at writing those, I think someone might rewrite this sometime.
As for the behavior, it's the same logic as `lerp_angle`, I'll do some tests tomorrow to see if i messed something up.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (AThousandShips) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:1235]: It looks like other methods use `Math.Abs` and so on, is that the standard approach here?

**Comment by adamscoble** [comment]
> As for the docs i am not the best at writing those, I think someone might rewrite this sometime. As for the behavior, it's the same logic as `lerp_angle`, I'll do some tests tomorrow to see if i messed something up.

In case it helps, I wrote my function docs (and names) to mirror Godot's, so feel free to pinch [them](https://github.com/adamscoble/godot-math-extensions/blob/main/addons/math_extensions/math_extensions.gd) if you'd like!

Also regarding @kleonc's comment, in my GDScript implementation the result of `move_toward_angle` is always `to`, which I believe is more intuitive than `to + TAU * x`.

**Comment by PrecisionRender** [comment]
@adamscoble I just straight-up copied the `lerp_angle` function code in my implementation.

````
static func move_toward_angle(from: float, to: float, delta: float) -> float:
	var difference = fmod(to - from, TAU)
	var distance = fmod(2.0 * difference, TAU) - difference
	return move_toward(from, from + distance, delta)
````

**Comment by adamscoble** [comment]
> @adamscoble I just straight-up copied the `lerp_angle` function code in my implementation.
Oh, well there you go! I don't use the lerp functions often for this stuff, as I created a [float smoothing class](https://github.com/adamscoble/godot-math-extensions/blob/main/addons/math_extensions/float_smooth.gd) in that same repository to avoid the framerate-dependent smoothing of lerp. So it seems you're being consistent with Godot!

**Comment by kleonc** [comment]
> Also regarding @kleonc's comment, in my GDScript implementation the result of `move_toward_angle` is always `to`, which I believe is more intuitive than `to + TAU * x`.

@adamscoble Your implementation behaves the same as the one in this PR, note the examples I gave in https://github.com/godotengine/godot/pull/69081#issuecomment-1601200040 are kinda edge cases (like why would you want to move away by 3600 degrees?). For both implementations if `delta < 0` the result is not being clamped whatsoever.

For the the `move_toward`/`Vector2.move_toward`/`Vector3.move_toward` it makes that `delta < 0` results in moving away from `to` without any clamping/limiting. The relevant 1D/2D/3D spaces are conceptually unlimited, meaning you can always move further away.
But "angle space" is different, it's periodic. You can't really move further away than 180 degrees from the given angle. So maybe `move_toward_angle` for `delta < 0` should be clamped/limited to `[to - PI, to + PI]` range (whether it should be wrapped to a range like `[0, TAU)` or `[-PI, PI)` is a different question).

As a simple example, what should be the result of:
```gdscript
move_toward_angle(deg_to_rad(90), deg_to_rad(0), deg_to_rad(-180))
```
Which can be interpreted as: we're at 90 degrees and we're moving 180 degrees away from 0 degrees.
Currently the result would be 270 degrees. I say maybe it should be clamped so the result would be 180 degrees (because when moving past 180 degrees we would start getting closer to the 0 degrees we were meant to be moving away from).

But maybe the current behavior is desired, I don't know. What I'm saying is that firstly the expected behavior should be well defined for all possible inputs. Tweaking the implementation/docs accordingly is the next step.

**Comment by PrecisionRender** [comment]
> But maybe the current behavior is desired, I don't know. What I'm saying is that firstly the expected behavior should be well defined for all possible inputs. Tweaking the implementation/docs accordingly is the next step.

I'd say it should work the same as `lerp_angle`, for the sake of consistency.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (ettiSurreal) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs]: Fixed!

**Comment by ettiSurreal** [comment]
Tested some things out in a GDScript port of the functions.
Clamping the value between [0.0, TAU]/[-PI, PI] seems to be as simple as using a wrap function on the return, I guess i should add that?
Negative values are seem to be an issue. Currently (if the wrap function is added),
> ```gdscript
> move_toward_angle(deg_to_rad(90), deg_to_rad(0), deg_to_rad(-180))
> ```
would return -90 degrees, since negative values currently do not support stopping at at the opposite angle. Which would mean if you run it per frame it will oscillate.
However, if you wanna move toward to the opposite angle you can just add 180 degrees to p_to and use a positive delta. I found a simple way to make it so the function does it automatically when you input a negative value, but it requires an If statement at the start of the function. Unity's equivalent function omits automatically doing it for optimization reasons, should we as well?

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:109]: The C# documentation is different from the GDScript documentation. We try to keep them in sync.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:111]:
```suggestion
        /// <param name ="to">The destination angle.</param>
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:124]:
```suggestion
        /// <param name ="to">The destination angle.</param>
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:1229]:
```suggestion
        /// <param name="to">The angle to move towards.</param>
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:1231]:
```suggestion
        /// <returns>The angle after moving.</returns>
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:1245]:
```suggestion
        /// <param name="to">The angle to move towards.</param>
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:1247]:
```suggestion
        /// <returns>The angle after moving.</returns>
```

**Review by raulsntos — CHANGES_REQUESTED** [review]

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (ettiSurreal) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs]: Fixed!

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs
Comment (ettiSurreal) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs]: Wrote a new doc entirely and copied it over to C#.

**Comment by ettiSurreal** [comment]
Wrapped the angle between [-PI, PI], since that's how `lerp_angle` seems to work.
For now decided to keep negative delta as "unsupported" for optimization (let me know if i should do otherwise), clarifying it in the docs, which I also just rewrote.

File: core/math/math_funcs.h
Comment (AThousandShips) [inline core/math/math_funcs.h:449]:
```suggestion
		return Math::wrapf(p_from + MIN(Math::abs(difference), p_delta) * SIGN(difference), -(float)Math_PI, (float)Math_PI);
```
We're explicit about the methods from this same class, also use the Godot specific minimum function not the global scope one

Same applies for the line above

Comment (AThousandShips) [reply core/math/math_funcs.h:449]: I'd also suggest prefixing our own functions here with `Math::` as that's partially adhered to

Comment (ettiSurreal) [reply core/math/math_funcs.h]: Done, I've seen a little bit of inconsistency with the former but that's probably for another PR.

**Comment by kleonc** [comment]
>Wrapped the angle between [-PI, PI], since that's how `lerp_angle` seems to work.

@ettiSurreal `lerp_angle` doesn't do any wrapping (and its behavior should remain unchanged as it would be a compatibility breaking change). Hence I think leaving `move_toward_angle` without wrapping the result to a specific range is the way to go. If desired, user can always wrap it manually, e.g. `wrapf(move_toward_angle(from, to, delta), -PI, PI)`.

>For now decided to keep negative delta as "unsupported" for optimization (let me know if i should do otherwise), clarifying it in the docs, which I also just rewrote.

Performance shouldn't differ much, I think the implementation would be as simple as:
```cpp
static _ALWAYS_INLINE_ double move_toward_angle(double p_from, double p_to, double p_delta) {
	double difference = angle_difference(p_from, p_to);
	double abs_difference = abs(difference);
	// When `delta < 0` move no further than to PI radians away from `p_to` (as PI is the max possible angle distance).
	return p_from + CLAMP(delta, abs_difference - Math_PI, abs_difference) * SIGN(difference);
}
```

---
 
> I'd say it should work the same as `lerp_angle`, for the sake of consistency.

@PrecisionRender Agree consistency is good but the thing is these already work differently as `lerp_angle` is not limiting the result in any direction. You can extrapolate by passing either `weight > 1` or `weight < -1`, e.g. `lerp_angle(0, PI, 99)` returns `99 * PI`.
To me `move_toward_angle` conceptually behaves differently, it's main purpose is to perform logic limiting the movement against the `to` parameter. Hence when moving away it makes sense to me to not move past 180 degrees away from `to`. So I'd limit it like that. But if others disagree and it's more desired to not limit moving away anyhow (possibly going full circles), it would be fine to me too. :upside_down_face:

**Comment by ettiSurreal** [comment]
> > Wrapped the angle between [-PI, PI], since that's how `lerp_angle` seems to work.
> 
> @ettiSurreal `lerp_angle` doesn't do any wrapping (and its behavior should remain unchanged as it would be a compatibility breaking change). Hence I think leaving `move_toward_angle` without wrapping the result to a specific range is the way to go. If desired, user can always wrap it manually, e.g. `wrapf(move_toward_angle(from, to, delta), -PI, PI)`.

I must've mistested it, I thought `lerp_angle` by itself stayed between [-PI, PI] unless you overloaded the weight.

> > For now decided to keep negative delta as "unsupported" for optimization (let me know if i should do otherwise), clarifying it in the docs, which I also just rewrote.
> 
> Performance shouldn't differ much, I think the implementation would be as simple as:
> 
> ```c++
> static _ALWAYS_INLINE_ double move_toward_angle(double p_from, double p_to, double p_delta) {
> 	double difference = angle_difference(p_from, p_to);
> 	double abs_difference = abs(difference);
> 	// When `delta < 0` move no further than to PI radians away from `p_to` (as PI is the max possible angle distance).
> 	return p_from + CLAMP(delta, abs_difference - Math_PI, abs_difference) * SIGN(difference);
> }
> ```
Since that's seemingly not a concern I can add your implementation then. My own implementation would've been something like this.
```c++
static _ALWAYS_INLINE_ double move_toward_angle(double p_from, double p_to, double p_delta) {
	if (p_delta > 0.0) {
		double difference = Math::angle_difference(p_from, p_to);
		return Math::wrapf(p_from + MIN(Math::abs(difference), p_delta) * SIGN(difference), -Math_PI, Math_PI);
	} else {
		double difference = Math::angle_difference(p_from, p_to + Math.PI);
		return Math::wrapf(p_from + MIN(Math::abs(difference), Math::abs(p_delta)) * SIGN(difference), -Math_PI, Math_PI);		
```

**Comment by PrecisionRender** [comment]
> @PrecisionRender Agree consistency is good but the thing is these already work differently as `lerp_angle` is not limiting the result in any direction. You can extrapolate by passing either `weight > 1` or `weight < -1`, e.g. `lerp_angle(0, PI, 99)` returns `99 * PI`.
> To me `move_toward_angle` conceptually behaves differently, it's main purpose is to perform logic limiting the movement against the `to` parameter. Hence when moving away it makes sense to me to not move past 180 degrees away from `to`. So I'd limit it like that. But if others disagree and it's more desired to not limit moving away anyhow (possibly going full circles), it would be fine to me too. 🙃

I've never run into a scenario where I need to do that moving _away_ from the desired angle, so I imagine both implementations would work.

**Comment by AncientSion** [comment]
can you add this to 3.5/6, i need it badly.

**Comment by AThousandShips** [comment]
@AncientSion Please don't bump PRs without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by AThousandShips** [comment]
This being a new feature it can be considered for cherry-picking, unsure if that will be done though

**Comment by timothyqiu** [comment]
Any reason why the function is named `move_toward_angle` instead of `rotate_toward`? I think it's technically not possible to _move_ toward an angle :P

File: doc/classes/@GlobalScope.xml
Comment (timothyqiu) [inline doc/classes/@GlobalScope.xml:80]:
```suggestion
				Returns the difference between the two angles. in the range of [code][-PI, +PI][/code]. When [param from] and [param to] are opposite, returns [code]-PI[/code] if [param from] is smaller than [param to], or [code]PI[/code] otherwise.
```

**Comment by PrecisionRender** [comment]
> Any reason why the function is named `move_toward_angle` instead of `rotate_toward`? I think it's technically not possible to _move_ toward an angle :P

I don't disagree, but I think these are the current reasons:

a. To be consistent with the `lerp_angle()` function
b. To be consistent with other engines

**Comment by timothyqiu** [comment]
> a. To be consistent with the `lerp_angle()` function

Make sense. But to be fair, "lerp" is one term (same goes for "cubic_interpolate"). The only choice for naming a rotation variant is adding a prefix or suffix :stuck_out_tongue: 

> b. To be consistent with other engines

Unity has inconsistent naming: the standalone version is named `MoveTowardsAngle`, while the member version is named `RotateTowards`.

**Comment by kleonc** [comment]
> Unity has inconsistent naming: the standalone version is named `MoveTowardsAngle`, while the member version is named `RotateTowards`.

And there's only `Vector3.RotateTowards`, no `Vector2.RotateTowards` or `Vector4.RotateTowards`.

Regarding the behavior for negative delta:
- [`MoveTowardsAngle`](https://docs.unity3d.com/ScriptReference/Mathf.MoveTowardsAngle.html):
>For optimization reasons, negative values of maxDelta are not supported and may cause oscillation. To push current away from a target angle, add 180 to that angle instead.
- [`Vector3.RotateTowards`](https://docs.unity3d.com/ScriptReference/Vector3.RotateTowards.html):
> If a negative value is used for maxRadiansDelta, the vector will rotate away from target/ until it is pointing in exactly the opposite direction, then stops.

So not only the naming is inconsistent. :upside_down_face:

---

Note that adding `Vector3.RotateTowards`-like method to `VectorN` types is also desired/requested, see: https://github.com/godotengine/godot-proposals/discussions/7302.

**Comment by Jamsers** [comment]
For anyone stumbling in here looking for an easy to use rotate_toward for both quats and basis' in the meantime, I used these:
```
func rotate_toward(from: Quaternion, to: Quaternion, delta: float) -> Quaternion:
	return from.slerp(to, min(1.0, delta / from.angle_to(to)))

func basis_rotate_toward(from: Basis, to: Basis, delta: float) -> Basis:
	return Basis(rotate_toward(from.get_rotation_quaternion(), to.get_rotation_quaternion(), delta))
```
Thanks to [kleonc](https://www.reddit.com/r/godot/comments/156gmx8/comment/jszze4p/) from Reddit for the quick fix ❤️

**Comment by ettiSurreal** [comment]
Okay i have no idea what to do with this I'll just make a new PR.

**Pull request closed without merging**
